internal/cli/cmd/cluster/artifact.go

```diff
@@ -102,24 +125,160 @@ func newArtifactDownloadCmd() *cobra.Command {
 			return err
 		}
-		defer reader.Close()
 
-		w, err := os.Create(dest)
+		fmt.Fprintf(console.Stdout(ctx), "Downloaded %s (namespace %s) to %s\n", src, namespace, dest)
+
+		return nil
+	})
+}
+
+func newArtifactCacheURLCmd() *cobra.Command {
+	var renew bool
+
+	return fncobra.Cmd(&cobra.Command{
+		Use:   "cache-url [target url] [destination filename] { --renew }",
+		Short: "Download an arbitrary URL using a pull-through cache.",
+		Long: `Download the content from an arbitrary URL and cache it for fast access.
+
+If the content is already present in the artifact cache for the given URL it will be used instead.
+
+The content at the URL is assumed to be immutable.`,
+		Args: cobra.ExactArgs(2),
+	}).WithFlags(func(flags *pflag.FlagSet) {
+		flags.BoolVar(&renew, "renew", false, "Force-download from the source and update the cached content.")
+	}).DoWithArgs(func(ctx context.Context, args []string) error {
+		sourceURL, dest := args[0], args[1]
+
+		parsedURL, err := url.Parse(sourceURL)
+		if err != nil {
+			return fnerrors.BadInputError("invalid URL format: %w", err)
+		}
+
+		token, err := auth.LoadDefaults()
+		if err != nil {
+			return fnerrors.BadDataError("failed to obtain auth data: %w", err)
+		}
+
+		cli, err := storage.NewClient(ctx, token)
 		if err != nil {
-			return fnerrors.New("failed to open %q: %w", dest, err)
+			return fnerrors.InvocationError("namespace api", "failed to connect: %w", err)
 		}
-		defer w.Close()
+		defer cli.Close()
 
-		if _, err := io.Copy(w, reader); err != nil {
-			return fnerrors.New("failed to write %q: %w", dest, err)
+		if !renew {
+			listResp, err := cli.Artifacts.ListArtifacts(ctx, &storagev1beta.ListArtifactsRequest{
+				Namespaces:  []string{cacheArtifactNamespace},
+				LabelFilter: []*stdlib.LabelFilterEntry{{Name: cacheSourceURLLabel, Value: sourceURL, Op: stdlib.LabelFilterEntry_EQUAL}},
+			})
+			if err != nil {
+				return fnerrors.InvocationError("namespace api", "failed to list artifacts: %w", err)
+			}
+			var newest *storagev1beta.Artifact
+			for _, art := range listResp.Artifacts {
+				if newest == nil || art.GetCreatedAt().AsTime().After(newest.GetCreatedAt().AsTime()) {
+					newest = art
+				}
+			}
+			if newest != nil {
+				fmt.Fprintf(console.Stderr(ctx), "Downloading from cache (cached at %v)...\n", newest.GetCreatedAt().AsTime())
+				if err := writeArtifact(ctx, cli, newest.GetNamespace(), newest.GetPath(), dest); err != nil {
+					return err
+				}
+				fmt.Fprintf(console.Stderr(ctx), "Downloaded to %s\n", dest)
+				return nil
+			} else {
+				fmt.Fprintf(console.Stderr(ctx), "Artifact not found in cache; downloading from source...\n")
+				// Fallthrough.
+			}
 		}
 
-		fmt.Fprintf(console.Stdout(ctx), "Downloaded %s (namespace %s) to %s", src, namespace, dest)
+		cachePath := cacheArtifactPath(time.Now(), parsedURL)
+		labels := map[string]string{cacheSourceURLLabel: sourceURL}
+
+		backoff.RetryNotify(func() error {
```

Comment:
I'd say that we are blindly retrying too many errors here.
I would invert the logic and treat all errors as permanent unless it is a known-safe-to-retry error.

Reply:
Done.